app/controllers/Application.java

```diff
@@ -1416,4 +1435,129 @@ public static Result restFlowMetricsGraphData(String flowDefId) {
   }
+
+  public static Result restDagGraphData(String flowExecId) {
+    int i,j,oneInnodeNum;
+    JsonObject datasets = new JsonObject();
+
+    if (flowExecId == null || flowExecId.isEmpty()) {
+      return ok(new Gson().toJson(datasets));
+    }
+
+    int max_index= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).findRowCount();
+    max_index= max_index+1;
+    int[][] adjMatrix = new int[max_index][max_index];
+
+    List<AppJobNameMap> reqList= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).findList();
+    for(AppJobNameMap temp1 : reqList){
+      int ownId= temp1.jobNameId;
+      String innodes= temp1.jobInnodes;
+      if(innodes!=null){
+        String[] innodeString = innodes.split(",");
+        for(String oneInnodeString : innodeString){
+          oneInnodeNum= Integer.parseInt(oneInnodeString);
+          adjMatrix[oneInnodeNum][ownId]= 1;
+        }
+      }
+    }
+
+    for (i = 1; i < max_index; i++) {
+      JsonObject row = new JsonObject();
+      for (j = 1; j < max_index; j++) {
+        if(adjMatrix[i][j]==1){
+          row.addProperty(Integer.toString(j), adjMatrix[i][j]);
+        }
+      }
+      AppJobNameMap tempStrStruct= AppJobNameMap.find.select("*").where().eq(AppJobNameMap.TABLE.FLOW_EXEC_ID, flowExecId).eq(AppJobNameMap.TABLE.JOB_NAME_ID, i).findUnique();
+      String label= tempStrStruct.jobName;
+
+      String colour1= "blue";
+
+      List<AppResult> timeStructList= AppResult.find.select("*").where().eq(AppResult.TABLE.FLOW_EXEC_ID, flowExecId).eq(AppResult.TABLE.JOB_NAME, label).findList();
+      long runningTime = 10;
+      if(timeStructList.isEmpty()){}else{
+        AppResult timeStruct= timeStructList.get(0);
+        if (timeStruct == null) {
+          // Dummy node
+        } else {
+          long sTime= timeStruct.startTime;
+          long fTime= timeStruct.finishTime;
+          runningTime= fTime - sTime;
+          runningTime= runningTime/5000;
+          logger.debug("Running time for is: " + runningTime);
+          label+= " (JOB TYPE :" + timeStruct.jobType + ") ";
+          colour1= "red";
+        }}
+      JsonObject dataset= new JsonObject();
+      dataset.add("row", row);
+      dataset.addProperty("label", label);
+      dataset.addProperty("time", runningTime);
+      dataset.addProperty("colour", colour1);
+      datasets.add(Integer.toString(i), dataset);
+    }
+
+    return ok(new Gson().toJson(datasets));
+  }
+
+
+  public static Result restMrDagGraphData(String jobExecId) {
+    JsonObject datasets = new JsonObject();
+
+    if (jobExecId == null || jobExecId.isEmpty()) {
+      return ok(new Gson().toJson(datasets));
+    }
+
+    int i, j;
+    List<AppResult> mrJobs= AppResult.find.select("*").where().eq(AppResult.TABLE.JOB_EXEC_ID, jobExecId).findList();
+    if(mrJobs.isEmpty()){
+      return null;
+    }
+    HashMap<String, Integer> jobToId= new HashMap<String, Integer>();
```

Comment:
I think the computation logic for calculating graph should not be here. We should have a separate class which should only return the computed adjecency matrix or adjacency list. We can also have an interface with the method that returns this list and which should be implemented by the jobtype classes. Something like GraphInterface with getGraphMatrix() interface which should be implemented by PigGraphData and HiveGraphData.